ci/lint.dart

```diff
@@ -0,0 +1,95 @@
+/// Runs clang-tidy on files with changes.
+///
+/// usage:
+/// dart lint.dart <path to compile_commands.json> <path to git repository> [clang-tidy checks]
+
+import 'dart:io' show File, Process, ProcessResult, exit;
+import 'dart:convert' show jsonDecode, utf8;
+
+class Command {
+  String directory;
+  String command;
+  String file;
+}
+
+Command parseCommand(Map<String, dynamic> map) {
+  return Command()
+    ..directory = map['directory']
+    ..command = map['command']
+    ..file = map['file'];
+}
+
+String calcTidyArgs(Command command) {
+  String result = command.command;
+  result = result.replaceAll(RegExp(r'\S*clang/bin/clang'), '');
+  result = result.replaceAll(RegExp(r'-MF \S*'), '');
+  return result;
+}
+
+String calcTidyPath(Command command) {
+  final RegExp regex = RegExp(r'\S*clang/bin/clang');
+  return regex
+      .stringMatch(command.command)
+      .replaceAll('clang/bin/clang', 'clang/bin/clang-tidy');
+}
+
+bool isNonEmptyString(String str) => str.length > 0;
+
+bool containsAny(String str, List<String> queries) {
+  for (String query in queries) {
+    if (str.contains(query)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Returns a list of all files with current changes or differ from `master`.
+List<String> getListOfChangedFiles(String repoPath) {
+  final Set<String> result = Set<String>();
+  final ProcessResult diffResult = Process.runSync(
+      'git', ['diff', '--name-only'],
+      workingDirectory: repoPath);
+  final ProcessResult diffCachedResult = Process.runSync(
+      'git', ['diff', '--cached', '--name-only'],
+      workingDirectory: repoPath);
+  final ProcessResult masterResult = Process.runSync(
+      'git', ['diff', '--name-only', 'master'],
+      workingDirectory: repoPath);
+  result.addAll(diffResult.stdout.split('\n').where(isNonEmptyString));
+  result.addAll(diffCachedResult.stdout.split('\n').where(isNonEmptyString));
+  result.addAll(masterResult.stdout.split('\n').where(isNonEmptyString));
+  return result.toList();
+}
+
+void main(List<String> arguments) async {
+  final String buildCommandsPath = arguments[0];
```

Comment:
Not sure what the state of the art in argument parsing in Dart is, but can there be a help string like with argparse in Python?

Reply:
You can use `package:args`, although the arguments used here are simple enough that I didn't really flag this. They could be even simpler if we just tell clang-tidy where to find compile_commands.json

Reply:
Yea, it's an external package, it complicates things.  I can add a help print if we aren't happy with the file comment.   No one should have this file as a binary without the source.